File: riddler/coffee_cards/coffee_cards.py
import numpy as np
import pandas as pd
import sys
import time


def make_trip(N=50):
    """
    Simulate random selection of coffee cards
    Each card starts with N drinks.
    Randomly pick a card until one of them runs out.
    When a card runs out what are the odds there are drinks
    left on the other card.

    Output = number of drinks remaining on other card
             when the selected card is exhausted
    """
    Na=N
    Nb=N
    while 1:
        if np.random.rand()<0.5:
            if Na==0:
                return(Nb)
            else:
                Na=Na-1
        else:
            if Nb==0:
                return(Na)
            else:
                Nb=Nb-1

if __name__ == "__main__":
    t1=time.time()
    www=[]
    num_trials=1000
    N=50
    if len(sys.argv)>1:
        num_trials=int(sys.argv[1])
    if len(sys.argv)>2:
        N=int(sys.argv[2])

    drinks=np.zeros(N+1,'i')
    for i in range(num_trials):
        k=make_trip(N)
        drinks[k]=drinks[k]+1
    counts=pd.DataFrame({'n':drinks})
    counts=counts.reset_index()
    counts.columns=['Drinks','Count']
    print("Remaining drink probability (starting with N = %d):" % N)
    print(counts)
    nzbins=counts.Drinks[counts.Count != 0].values
    print("Min Remaining = ",nzbins[0])
    print("Max Remaining = ",nzbins[-1])
    print("Mean remaining = ",sum(counts.Drinks*counts.Count)/sum(counts.Count))
    t2=time.time()
    print("Execution time (seconds) = ",t2-t1)

File: riddler/coffee_cards/coffee_cards.lua
make_trip = function(n)
    n=n or 50
    na=n
    nb=n
    while true do
        if math.random()<0.5 then
            if na==0 then
                return(nb)
            else
                na=na-1
            end
        else
            if nb==0 then
                return(na)
            else
                nb=nb-1
            end
        end
    end
end


num_trials = arg[1] or 1000
num_drinks = arg[2] or 50

n={}
for i=0,num_drinks do n[i]=0 end

t1=os.time()
for trial=1,num_trials do
    k=make_trip()
    n[k]=n[k]+1
end
t2=os.time()

print("Execution time " .. t2-t1)

nmax=0
for i=0,num_drinks do
    if n[i]>0 then
        nmax=i
    end
end

s=0
for i=0,num_drinks do s=s+i*n[i]/num_trials end

print("Number of trials = " .. num_trials)
print("Drinks per card  = " .. num_drinks)
print("Mean drinks remaining = " .. s)
print("Max  drinks remaining = " .. nmax)

for i=0,nmax do print(i,n[i]) end
